contracts/account/extensions/ERC7821.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {ERC7579Utils, Mode, CallType, ExecType, ModeSelector} from "@openzeppelin/contracts/account/utils/draft-ERC7579Utils.sol";
+import {IERC7821} from "../../interfaces/IERC7821.sol";
+import {AccountCore} from "../AccountCore.sol";
+
+/**
+ * @dev Minimal batch executor following ERC-7821. Only supports basic mode (no optional "opData").
+ */
+abstract contract ERC7821 is IERC7821 {
+    using ERC7579Utils for *;
+
+    error UnsupportedExecutionMode();
+
+    /**
+     * @dev Executes the calls in `executionData` with no optional `opData` support.
+     *
+     * NOTE: Access to this function is controlled by {_erc7821AuthorizedExecutor}. Changing access permissions, for
+     * example to approve calls by the ERC-4337 entrypoint, should be implement by overriding it.
+     *
+     * Reverts and bubbles up error if any call fails.
+     */
+    function execute(bytes32 mode, bytes calldata executionData) public payable virtual {
+        if (!_erc7821AuthorizedExecutor(mode, executionData)) revert AccountCore.AccountUnauthorized(msg.sender);
+        if (!supportsExecutionMode(mode)) revert UnsupportedExecutionMode();
+        executionData.execBatch(ERC7579Utils.EXECTYPE_DEFAULT);
+    }
+
+    /// @inheritdoc IERC7821
+    function supportsExecutionMode(bytes32 mode) public view virtual returns (bool result) {
+        (CallType callType, ExecType execType, ModeSelector modeSelector, ) = Mode.wrap(mode).decodeMode();
+        return
+            callType == ERC7579Utils.CALLTYPE_BATCH &&
+            execType == ERC7579Utils.EXECTYPE_DEFAULT &&
+            modeSelector == ModeSelector.wrap(0x00000000);
+    }
+
+    /**
+     * @dev Access control mechanism for the {execute} function.
+     */
+    function _erc7821AuthorizedExecutor(
+        bytes32 /* mode */,
+        bytes calldata /* executionData */
+    ) internal view virtual returns (bool) {
```

Comment:
any reason to to remove `address caller` from the parameter, and using `msg.sender` implicitelly. I have the feeling its nicer if this internal function can be called to check the permission of someone else.

Reply:
I wanted to make this function composable with `onlyOwner`, `onlyRole` or `restricted` modifiers in the same way `authorizeUpgrade` is in UUPSUpgradeable. I start to think that the `AccessManaged` contract must expose a variant of the `restricted` modifier that allow passing in a custom address. Also note those modifiers generally revert and not return a boolean

Reply:
Worth noting that if we want this function to be composable with `AccessManaged.restricted`, then it can't have `view` visibility :(